src/py/flwr/common/config_test.py

```diff
@@ -0,0 +1,164 @@
+# limitations under the License.
+# ==============================================================================
+"""Test util functions handling Flower config."""
+
+import os
+import textwrap
+from pathlib import Path
+from unittest.mock import patch
+
+import pytest
+
+from .config import (
+    flatten_dict,
+    get_flwr_dir,
+    get_project_config,
+    get_project_dir,
+    parse_config_args,
+)
+
+# Mock constants
+FAB_CONFIG_FILE = "pyproject.toml"
+
+
+def test_get_flwr_dir_with_provided_path() -> None:
+    """Test get_flwr_dir with a provided valid path."""
+    provided_path = "."
+    assert get_flwr_dir(provided_path) == Path(provided_path).absolute()
+
+
+def test_get_flwr_dir_without_provided_path() -> None:
+    """Test get_flwr_dir without a provided path, using default home directory."""
+    with patch.dict(os.environ, {"HOME": "/home/user"}):
+        assert get_flwr_dir() == Path("/home/user/.flwr")
+
+
+def test_get_flwr_dir_with_flwr_home() -> None:
+    """Test get_flwr_dir with FLWR_HOME environment variable set."""
+    with patch.dict(os.environ, {"FLWR_HOME": "/custom/flwr/home"}):
+        assert get_flwr_dir() == Path("/custom/flwr/home")
+
+
+def test_get_flwr_dir_with_xdg_data_home() -> None:
+    """Test get_flwr_dir with FLWR_HOME environment variable set."""
+    with patch.dict(os.environ, {"XDG_DATA_HOME": "/custom/data/home"}):
+        assert get_flwr_dir() == Path("/custom/data/home/.flwr")
+
+
+def test_get_project_dir_invalid_fab_id() -> None:
+    """Test get_project_dir with an invalid fab_id."""
+    with pytest.raises(ValueError):
+        get_project_dir("invalid_fab_id", "1.0.0")
+
+
+def test_get_project_dir_valid() -> None:
+    """Test get_project_dir with an valid fab_id and version."""
+    app_path = get_project_dir("app_name/user", "1.0.0", flwr_dir=".")
+    assert app_path == Path("apps") / "app_name" / "user" / "1.0.0"
+
+
+def test_get_project_config_file_not_found() -> None:
+    """Test get_project_config when the configuration file is not found."""
+    with pytest.raises(FileNotFoundError):
+        get_project_config("/invalid/dir")
+
+
+def test_get_project_config_file_valid(tmp_path: Path) -> None:
+    """Test get_project_config when the configuration file is not found."""
+    pyproject_toml_content = """
+        [build-system]
+        requires = ["hatchling"]
+        build-backend = "hatchling.build"
+
+        [project]
+        name = "fedgpt"
+        version = "1.0.0"
+        description = ""
+        license = {text = "Apache License (2.0)"}
+        dependencies = [
+            "flwr[simulation]>=1.9.0,<2.0",
+            "numpy>=1.21.0",
+        ]
+
+        [flower]
+        publisher = "flwrlabs"
+
+        [flower.components]
+        serverapp = "fedgpt.server:app"
+        clientapp = "fedgpt.client:app"
+
+        [flower.engine]
+        name = "simulation" # optional
+
+        [flower.engine.simulation.supernode]
+        count = 10 # optional
```

Comment:
These should start with `flower.config`

Reply:
@danieljanes Well I've added tests for this function because it didn't have any before, but this is not really the scope of the PR. So this function doesn't do anything particular with `flower.config`

Reply:
I updated it